cmd/avsc2avdl/main.go

```diff
@@ -0,0 +1,345 @@
+	}
+	if flag.Parse(os.Args[1:]) != nil {
+		return 2
+	}
+	if flag.NArg() != 1 {
+		flag.Usage()
+		return 2
+	}
+	if err := avsc2avdl(flag.Arg(0)); err != nil {
+		fmt.Fprintf(os.Stderr, "avsc2avdl: %v\n", err)
+		return 1
+	}
+	return 0
+}
+
+func avsc2avdl(avscFile string) error {
+	data, err := ioutil.ReadFile(avscFile)
+	if err != nil {
+		return err
+	}
+	at, err := typeinfo.ParseSchema(string(data), nil)
+	if err != nil {
+		return fmt.Errorf("cannot parse schema from %q: %v", avscFile, err)
+	}
+	ref, ok := at.(*schema.Reference)
+	if !ok {
+		return fmt.Errorf("top level of schema is not a reference")
+	}
+	g := &generator{
+		filename: avscFile,
+		done:     make(map[schema.QualifiedName]bool),
+	}
+	g.addDefinition(ref.Def)
+	// Use the top level definition's namespace as the default namespace for
+	// all definitions.
+	namespace := ref.Def.AvroName().Namespace
+	g.pushNamespace(namespace)
+	if namespace != "" {
+		g.printf("@namespace(%q)\n", namespace)
+	}
+	g.printf("protocol _ {\n")
+	for i := 0; ; i++ {
+		def := g.removeDefinition()
+		if def == nil {
+			break
+		}
+		if i > 0 {
+			g.printf("\n")
+		}
+		g.writeDefinition(def)
+	}
+	g.printf("}\n")
+	os.Stdout.Write(g.buf.Bytes())
+	return nil
+}
+
+type generator struct {
+	filename       string
+	buf            bytes.Buffer
+	queue          []schema.Definition
+	namespaceStack []string
+	done           map[schema.QualifiedName]bool
+}
+
+func (g *generator) printf(f string, a ...interface{}) {
+	fmt.Fprintf(&g.buf, f, a...)
+}
+
+func (g *generator) writeDefinition(def schema.Definition) {
+	name := def.AvroName()
+	writeNamespace := func() {}
+	if name.Namespace != "" && name.Namespace != g.namespace() {
+		writeNamespace = func() {
+			g.printf("\t@namespace(%q)\n", name.Namespace)
+		}
+		g.pushNamespace(name.Namespace)
+		defer g.popNamespace()
+	}
+	switch def := def.(type) {
+	case *schema.RecordDefinition:
+		g.writeMetadata(def, "\t")
+		writeNamespace()
+		g.printf("\trecord %s {\n", name.Name)
+		for _, field := range def.Fields() {
+			g.writeMetadata(field, "\t\t")
+			g.printf("\t\t%s %s", g.typeString(field.Type()), field.Name())
+			if field.HasDefault() {
+				g.printf(" = %s", jsonMarshal(field.Default(), "\t\t"))
+				if isEnum(field.Type()) {
+					fmt.Fprintf(os.Stderr, "%s: WARNING: default value (%#v) for enum-valued field in %s.%s will be ignored (see https://issues.apache.org/jira/browse/AVRO-2866)\n",
+						g.filename,
+						field.Default(),
+						name,
+						field.Name(),
+					)
+				}
+			}
+			g.printf(";\n")
+		}
+		g.printf("\t}\n")
+	case *schema.EnumDefinition:
+		g.writeMetadata(def, "\t")
+		writeNamespace()
+		g.printf("\tenum %s {\n", name.Name)
+		syms := def.Symbols()
+		for i, sym := range syms {
+			g.printf("\t\t%s", sym)
+			if i < len(syms)-1 {
+				g.printf(",")
+			}
+			g.printf("\n")
+		}
+		// TODO support enum defaults (gogen-avro doesn't make it easy to get
+		// them - it might not support them currently itself)
```

Comment:
We should print this warning to `stderr` too

Reply:
yeah, but it's not easy to get the default value and we don't use enum defaults yet, so i think i'll leave it for now.